spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/handler/AsyncPredicate.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2013-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.springframework.cloud.gateway.handler;
+
+import java.util.Objects;
+import java.util.function.Function;
+
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+/**
+ * @author Ben Hale
+ */
+public interface AsyncPredicate<T> extends Function<T, Publisher<Boolean>> {
```

Comment:
why publisher and not `CompletionStage`?

Reply:
Isn't `public interface AsyncPredicate<T> extends Function<T, Mono<Boolean>>` a simpler signature. Just tested this locally and it appears to work with small changes to only this file

Reply:
Use of `Publisher` is idiotmatic in this case. Putting `Mono` in the interface limits all implementors to using Reactor, preventing the use of any other Reactive Streams compliant implementation (e.g. RxJava). Declaring`Publisher` results in the greatest flexibility, with no penalty in performance since Reactor short-circuits using a cast, when other Reactor types are encountered.